### PR TITLE
Migrate Joined flow part 1

### PR DESCRIPTION
**Pull request by rpdome**
Mostly 1:1 migration, except the following.

1. Extracts OAuth2ErrorCode into its own class (will do the same for SubError and other classes in `AuthenticationConstants` that we haven't started migrating)
2. Merge `ILocalAuthenticationResultBase` into `ILocalAuthenticationResult`.

Also get rid of some duplicated functions/constants.

### CODE REVIEW

**Review by AdamBJohnsonx — APPROVED** [review]
This looks straightforward.  I think we're in for a fun merge.

File: common4j/src/main/com/microsoft/identity/common/java/util/StringUtil.java [hunk 1]
```diff
@@ -351,4 +356,38 @@ public static String sanitizeNullAndLowercaseAndTrim(final String input) {
 
         return outValue;
     }
+
+    /**
+     * encode string with url form encoding. Space will be +.
```
Comment (shahzaibj) [inline common4j/src/main/com/microsoft/identity/common/java/util/StringUtil.java:361]:
```suggestion
     * Encode string with url form encoding. Space will be +.
```

Comment (shahzaibj) [reply common4j/src/main/com/microsoft/identity/common/java/util/StringUtil.java:361]: "Space will be plus dot" ~ What does this mean?

File: common4j/src/main/com/microsoft/identity/common/java/util/StringUtil.java [hunk 2]
```diff
@@ -351,4 +356,38 @@ public static String sanitizeNullAndLowercaseAndTrim(final String input) {
...
+     *
+     * @param source the string to encode.
+     * @return the decoded
```
Comment (shahzaibj) [inline common4j/src/main/com/microsoft/identity/common/java/util/StringUtil.java:364]:
```suggestion
     * @return the decoded String
```

File: common4j/src/main/com/microsoft/identity/common/java/util/StringUtil.java [hunk 3]
```diff
@@ -351,4 +356,38 @@ public static String sanitizeNullAndLowercaseAndTrim(final String input) {
...
+    }
+
+    /**
+     * Joins the string in the given set into a single string with the provided delimiter.
+     *
+     * @param delimiter     the delimiter string.
+     * @param stringsToJoin set of strings to join.
+     */
+    public static String join(@NonNull final String delimiter,
+                              @NonNull final Set<String> stringsToJoin) {
```
Comment (shahzaibj) [inline common4j/src/main/com/microsoft/identity/common/java/util/StringUtil.java:378]: If we move to java 8 then we can just do String.join(delimeter, String args...)

**Review by shahzaibj — APPROVED** [review]

File: common4j/src/main/com/microsoft/identity/common/java/util/StringUtil.java (hunk 1 above)
Comment (rpdome) [reply common4j/src/main/com/microsoft/identity/common/java/util/StringUtil.java:362]: @shahzaibj I think it meant "space will be replaced by '+'". dot is just a fullstop.

**Review by iambmelt — APPROVED** [review]
I don't see any issues

**Pull request merged**